egg/core/language_analysis.py

```diff
@@ -260,4 +260,34 @@ def on_epoch_end(self, _loss, logs: Interaction, epoch: int):
 
     def on_test_end(self, loss, logs, epoch):
         if self.print_test:
-            self.print_message(logs, "test", epoch)
+            self.print_message(logs, 'test', epoch)
+
+# The PrintValidationEvent callback function checks that we are at the
+# last epoch (either the last epoch required by the user, or because
+# of early stopping), and it prints sender_input, labels, message and
+# receiver_output for all data points in the validation set.
+# These data are stored in an Interaction object (see interaction.py
+# under core), that logs various data about each game data point.
+class PrintValidationEvents(Callback):
+    def __init__(self, n_epochs):
+        super().__init__()
+        self.n_epochs=n_epochs
+
+    @staticmethod
+    def print_events(logs: Interaction):
+        print("INPUTS")
+        print([m.tolist() for m in logs.sender_input], sep='\n')
+        print("LABELS")
+        print([m.tolist() for m in logs.labels], sep='\n')
+        print("MESSAGES")
+        print([m.tolist() for m in logs.message], sep='\n')
+        print("OUTPUTS")
+        print([m.tolist() for m in logs.receiver_output], sep='\n')
+        
+    def on_test_end(self, _loss, logs: Interaction, epoch: int): # here is where we make sure we are printing the validation set (on_test_end, not on_epoch_end)
+        if (epoch==self.n_epochs): # here is where we check that we are at the last epoch
+            self.print_events(logs)
+
+    def on_early_stopping(self,  _train_loss, _train_logs, epoch, _test_loss, test_logs): # same behaviour if we reached early stopping
+        self.print_events(test_logs)
```

Comment:
Was this tested? I thing the signature should be:

```
def on_early_stopping(
        self,
        train_loss: float,
        train_logs: Interaction,
        epoch: int,
        test_loss: float = None,
        test_logs: Interaction = None,
    ):
```

Reply:
I have tested it and it works. And indeed it seems to me to have the correct arguments: where do you see a discrepancy?